Implement Type->looseCompare(type)

staabm opened the pull request:
closes https://github.com/phpstan/phpstan/issues/8224

staabm committed:
Implement Type->looseCompare(type)

staabm commented:
for now I have implemented the `looseCompare` methods for all types.
now I need to have a look at the failling tests and afterwards type narrowing

ondrejmirtes commented:
Type narrowing is a separate concern, I wouldn't do that here.

ondrejmirtes commented on src/Reflection/InitializerExprTypeResolver.php:
```diff
@@ -1354,7 +1354,7 @@ public function resolveEqualType(Type $leftType, Type $rightType): BooleanType
 		}
 
-		return new BooleanType();
+		return $leftType->looseCompare($rightType);
```
This whole method needs to be just `$leftType->looseCompare($rightType)`.

staabm replied:
In case we want to move all the logic we have here into `$leftType->looseCompare($rightType)` we would need to pass the PhpVersion as an additional param, right?
(Loose comparison differs between php versions)

ondrejmirtes commented on src/Type/Constant/ConstantStringType.php:
```diff
@@ -278,6 +278,16 @@ public function toArrayKey(): Type
+
+	public function isFloat(): TrinaryLogic
+	{
+		return TrinaryLogic::createMaybe();
```
These additions are wrong. ConstantStringType can never be an integer, nor a float.

ondrejmirtes reviewed (changes requested):
I don't see the table of `==` comparison described anywhere in this PR https://www.php.net/manual/en/types.comparisons.php.

All types need to take into account being compared to any other type. For example this implementation in NullType is too naive:

```php
return $type->isNull()->toBooleanType();
```

Because all of the following examples are true:

```php
null == []
null == false
null == 0
null == ""
```

ondrejmirtes commented:
Also an idea for a preliminary or a future PR - `NanType`. See the last comment here: https://www.php.net/manual/en/function.is-nan.php

staabm committed:
fix types

staabm committed:
more tests

staabm commented:
I have implemented a few basic types and added tests for them in `loose-compare.php`. please have a look whether this is done how you imagined it.

btw: this PR will get really big if we implement all combinations for all types here.
do you think we could split this, e.g. doing basic and simple object type in one PR, and all more specialized types in separate PRs?

thank you

ondrejmirtes commented on src/Type/Traits/ConstantScalarTypeTrait.php:
```diff
@@ -46,6 +49,29 @@ public function isSuperTypeOf(Type $type): TrinaryLogic
+
+		if ($type instanceof ConstantScalarType) {
+			// @phpstan-ignore-next-line
+			return new ConstantBooleanType($this->getValue() == $type->getValue()); // phpcs:ignore
```
I'd prefer to implement logic instead of letting PHP runtime do the operation. Because when it's different across PHP versions, it means that the result on the playground wouldn't be representative.

Why? Playground runs on PHP 8.1, but shows results PHP 7.2-PHP8.2. So if `==` between scalars results in different boolean on different version, the playground wouldn't show it.

And there are definitely some differences. A friend recently gave a talk about some gotchas, I asked him for the slides.

herndlm replied:
I think I triggered one oft those gotchas last week resulting in a very serious bug https://3v4l.org/MEjDC

staabm replied:
nice catch. lets do this first as a [decoupled PR](https://github.com/phpstan/phpstan-src/pull/2220)

ondrejmirtes commented on src/Type/IntegerType.php:
```diff
@@ -128,6 +129,22 @@ public function isScalar(): TrinaryLogic
+			if ($stringType->isNumericString()->yes()) {
+				continue;
+			}
+			return new ConstantBooleanType(false);
```
I think this is wrong. So you get a union like `'a'|'123'`, it doesn't have to be false.

ondrejmirtes reviewed (changes requested):
I like StringType::looseCompare but otherwise it's still gonna be a lot of work 😅 And we're not even diving into type narrowing now. I think it's obvious why I was avoiding this topic until now 😂

ondrejmirtes commented:
FYI you might find a lot of wisdom and test cases here in these slides from @janedbal (not everything is about `==`): 
[Is-PHP-still-fucked-up_.pdf](https://github.com/phpstan/phpstan-src/files/10663056/Is-PHP-still-fucked-up_.pdf)


We might also need to do NanType and InfType because of these...

staabm committed:
Restore src/Reflection/InitializerExprTypeResolver.php

staabm committed:
Restore src/Type/Accessory/AccessoryArrayListType.php

staabm committed:
Restore src/Type/Accessory/AccessoryNonEmptyStringType.php

staabm committed:
Restore src/Type/Accessory/AccessoryNonFalsyStringType.php

staabm committed:
Restore src/Type/Accessory/AccessoryNumericStringType.php

staabm committed:
Restore src/Type/Accessory/HasOffsetType.php

staabm committed:
Restore src/Type/Accessory/HasOffsetValueType.php

staabm committed:
Restore src/Type/Accessory/NonEmptyArrayType.php

staabm committed:
Restore src/Type/Accessory/OversizedArrayType.php

staabm committed:
Fix

staabm committed:
Update InitializerExprTypeResolver.php

staabm committed:
test never

staabm committed:
loose array tests

staabm committed:
Update loose-comparisons.php

staabm committed:
fix

staabm committed:
Restore src/Type/IterableType.php

staabm committed:
Update loose-comparisons.php

staabm committed:
Restore src/Type/ResourceType.php

staabm committed:
Update loose-comparisons.php

staabm committed:
fix

staabm committed:
fix

staabm commented:
I added tons of tests and implemented `looseCompare` on a lot of types.

I happy for everyone giving feedback, since there are a lot of test-cases added and its pretty likely I did not assert everything right :)

ondrejmirtes commented on src/Type/Accessory/AccessoryNonEmptyStringType.php:
```diff
@@ -262,6 +264,12 @@ public function isScalar(): TrinaryLogic
 	{
+		$looseFalse = new NullType();
+
+		if ($looseFalse->isSuperTypeOf($type)->yes()) {
```
We can be sure it's going to be always false with:

* `''`
* `false`

ondrejmirtes replied:
I see that this test passes but I have no idea how:

```
assertType('false', $nonEmptyString == $emptyStr);
```

staabm replied:
> I see that this test passes but I have no idea how

it works because it is resolved via the "identical" type in https://github.com/phpstan/phpstan-src/blob/d01b91999311fbcdc091e105ac46df536dcc5648/src/Reflection/InitializerExprTypeResolver.php#L1330-L1337

it will start failing as expected, after https://github.com/phpstan/phpstan-src/pull/2228 is merged :)

staabm replied:
`'0' == false` is `true`

ondrejmirtes commented on src/Type/ArrayType.php:
```diff
@@ -348,6 +348,26 @@ public function isScalar(): TrinaryLogic
+		$looseTrue = new ConstantBooleanType(true);
+
+		if ($looseTrue->isSuperTypeOf($type)->yes()) {
+			return new ConstantBooleanType(true);
```
I don't think so. Array might be empty, and `[] == true` is false.

ondrejmirtes commented on src/Type/ArrayType.php:
```diff
@@ -348,6 +348,26 @@ public function isScalar(): TrinaryLogic
+
+		$looseFalse = new UnionType([
+			new ObjectWithoutClassType(),
+			new ConstantBooleanType(false),
```
Array can be equal to `false` if it's empty.

ondrejmirtes commented on src/Type/BooleanType.php:
```diff
@@ -121,6 +121,10 @@ public function isScalar(): TrinaryLogic
 	public function looseCompare(Type $type, PhpVersion $phpVersion): BooleanType
 	{
+		if ($type->isObject()->yes()) {
+			return new ConstantBooleanType(true);
```
`false` is equal with any object? I don't think so.

ondrejmirtes reviewed (changes requested):
`InitializerExprTypeResolver::resolveEqualType` should be just `return $leftType->looseCompare($rightType, $this->phpVersion);`, nothing more.

staabm committed:
Review fixes

staabm commented on src/Type/ArrayType.php:
```diff
@@ -348,6 +348,19 @@ public function isScalar(): TrinaryLogic
+			new NullType(),
+			new IntegerType(),
+			new FloatType(),
+			new NeverType(),
```
should all type turn a possible `*NEVER*` operand into a `false`, or just leave it a undecided boolean?

staabm commented on tests/PHPStan/Analyser/data/loose-comparisons.php:
```diff
@@ -518,5 +600,881 @@ public function sayEmptyStr(
+		assertType('false', $callable == $zero);
+		assertType('false', $callable == 10);
+		assertType('false', $callable == $minusOne);
+		assertType('bool', $callable == $oneStr); // could be false, because invalid function name
+		assertType('bool', $callable == $zeroStr); // could be false, because invalid function name
+		assertType('bool', $callable == $minusOneStr); // could be false, because invalid function name
```
within the type system we cannot use reflection to check whether given strings are valid callables.
we could do some kind of regular expression matching, whether a constant string contains only of chars allowed within a function name, but I don't think its worth it

staabm committed:
list type

staabm committed:
non falsey string

staabm committed:
numeric strings

staabm committed:
add class-string type

staabm commented:
ondrej, do you agree that we can implement only the most common types (all scalars, arrays, basic object types) with this PR, while leaving more specialized ones for a future PR?

staabm committed:
fix deprecations

staabm committed:
compare class-strings to non-class-strings

ondrejmirtes commented on src/Type/Accessory/AccessoryNumericStringType.php:
```diff
@@ -275,6 +276,10 @@ public function isScalar(): TrinaryLogic
 	public function looseCompare(Type $type, PhpVersion $phpVersion): BooleanType
 	{
+		if ($type->isString()->yes() && $type->isNumericString()->no()) {
+			return new ConstantBooleanType(false);
```
This one depends on the PhpVersion, right? Because of https://wiki.php.net/rfc/string_to_number_comparison

staabm replied:
I don't think so, since we compare string with string here.

The rfc describes phpversion specific differences in comparing non-numeric string and numbers

ondrejmirtes commented on src/Type/Accessory/NonEmptyArrayType.php:
```diff
@@ -354,6 +355,18 @@ public function isScalar(): TrinaryLogic
+			return new ConstantBooleanType(true);
+		}
+
+		if ($type->isFalse()->yes() || $type->isNull()->yes()) {
```
This kind of condition fails for `false|null`, because both left and right part are maybe. You need to use (new UnionType(false, null))->isSuperTypeOf($type)->yes()`.

You also need to include the below `if` in that union type too - an empty ConstantArrayType.

staabm replied:
fixed

ondrejmirtes commented on src/Type/IntegerType.php:
```diff
@@ -131,6 +133,20 @@ public function isScalar(): TrinaryLogic
 	{
+		if (count($type->getConstantStrings()) > 0) {
+			foreach ($type->getConstantStrings() as $stringType) {
+				if ($stringType->isNumericString()->yes()) {
```
Depends on PhpVersion, right?

staabm replied:
right, fixed

ondrejmirtes reviewed (changes requested):
I just looked at several implementations, not all of them. Also please limit your pushes, the email conversation is unbearably big in my email client, there's no reason you needed 74 force-pushes. Push only when the build is green for you locally and you want to show me your code. Thank you.

staabm committed:
fix falsy union types

staabm committed:
fix enums

staabm committed:
fix build

staabm committed:
fix

staabm committed:
fix php7 expectations

staabm committed:
de-duplicate zero comparison
